pallets/collator-allowlist/src/lib.rs

```diff
@@ -0,0 +1,139 @@
+//! have chosen the right staking mechanism.
+#![cfg_attr(not(feature = "std"), no_std)]
+use frame_support::{dispatch::DispatchResult, traits::ValidatorRegistration};
+
+use frame_system::ensure_root;
+
+pub use pallet::*;
+
+pub mod weights;
+pub use weights::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	// Import various types used to declare pallet in scope.
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	// Simple declaration of the `Pallet` type. It is placeholder we use to implement traits and
+	// method.
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_session::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Type representing the weight of this pallet
+		type WeightInfo: WeightInfo;
+	}
+
+	// The genesis config type.
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub initial_state: Vec<T::ValidatorId>,
+	}
+
+	// The default value for the genesis config type.
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				initial_state: vec![],
+			}
+		}
+	}
+
+	// The build of genesis for the pallet.
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			self.initial_state
+				.iter()
+				.for_each(|id| <Allowlist<T>>::insert(id, ()));
+		}
+	}
+
+	/// The collator's allowlist.
+	/// Note: We implement it as a close-enough HashSet: Map<ValidatorId, ()>.
+	#[pallet::storage]
+	pub(super) type Allowlist<T: Config> = StorageMap<_, Blake2_256, T::ValidatorId, ()>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A collator has been added to the allowlist
+		CollatorAdded(T::ValidatorId),
+
+		/// A collator has been removed from the allowlist
+		CollatorRemoved(T::ValidatorId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The collator has already been added to the allowlist.
+		CollatorAlreadyAllowed,
+
+		/// The provided collator was not found in the storage.
+		CollatorNotPresent,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add the given `collator_id` to the allowlist.
+		/// Fails if
+		///   - `origin` fails the `ensure_root` check
+		///   - `collator_id` is already part of the allowlist
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::allow())]
+		pub fn add(origin: OriginFor<T>, collator_id: T::ValidatorId) -> DispatchResult {
+			ensure_root(origin)?;
+
+			ensure!(
+				!<Allowlist<T>>::contains_key(&collator_id),
+				Error::<T>::CollatorAlreadyAllowed
+			);
+			<Allowlist<T>>::insert(collator_id.clone(), ());
+			Self::deposit_event(Event::CollatorAdded(collator_id));
+			Ok(())
+		}
```

Comment:
We should add a check here, that checks if the collator has already uploaded its keys to the session pallet. 
Maybe via wiring in the Session pallet as a `ValidatorRegistration` type.

Reply:
Hm. What if the collator had uploaded its keys to the session pallet, we verify that here and thus allowlist them, and afterwards they unload their keys?

Point being, if we don't have a way to reactively add and remove collators from the allow list based on the state of their keys in the session pallet, it's not very meaningful to add that check here.

Reply:
I like the idea of duplicating a bit the session key check. The main reason is that we enforce whoever we want to allow as collator to have the session keys before whitelisting (and we do not need to do any manual check). 
As @NunoAlexandre said they could remove the keys in between, but at least we have an initial check that would hopefully remove 90% of the readiness of a collator when onboarding.
In the case that they do remove keys in between whitelisting and registering, registering will fail (assuming empty `0x` is not a successful `isSome`). And in the case that they remove the keys after registration, then collator will not produce blocks in next session, and every session change the collator-selection logic will kick out those that did not produce blocks.
We of course keep the session key check as well in the `is_registered`.

Reply:
Ok, that sounds reasonable to me and put in practise it will be handy, indeed. I will check whether there is a way to hook on `purge_keys` or so to keep this storage in sync with the `session` pallet, but that is something we can do afterwards.